### ISSUE TITLE
Does Llava-Pythia-1.3B have visual question answering ability?

### ISSUE THREAD

**Issue by NextFutureQin** [liyaxuanliyaxuan/TinyVLA#15]
**We made corresponding calls and tried to find that the output effect was not good**

用户: how can i use knife
助手: BethBre deservehubhubhub albumin Binding Mex cockodia Bible Appro alwayshub Terryeter viable Bachophageal SM instance HER instance populated bias murmured murmured unfamiliar unfamiliar unfamiliar unfamiliar unfamiliar unfamiliar comics与 chat superureth playful tex indicate 24 chemok naturrecogn generality enablingconnected enabling bread nex allows enabling enablingscreen allowing allowing allowing expectOn populatedamong spacious pus nodding unter topped sä siempreaporeognами enabling implementedshore bour encouraging sexually controlled venues environmenteffectilateralognами 66 votre donc donc donc donc donc donc donc movement umbrella heated nicer arrangement arrangements arrangements arrangements arrangements arrangements Immediately Ventless TER TER Conse vict dir donc ESP phenomenological phenomenological humid chemistry GChigheramong canvas connects GOP GOP phenomenological connects

**demo.py:**
`
import torch
from transformers import AutoTokenizer, AutoConfig
from llava_pythia01 import conversation as conversation_lib
from llava_pythia01.model.language_model.pythia.llava_pythia import LlavaPythiaForCausalLM

def _get_output_embeddings(self):
    # 强制把 token embedding 当作输出 head
    return self.gpt_neox.embed_in
LlavaPythiaForCausalLM.get_output_embeddings = _get_output_embeddings

MODEL_DIR = "/root/TinyVLA/Llava-Pythia-1.3B"

config = AutoConfig.from_pretrained(MODEL_DIR, trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained(
    MODEL_DIR,
    trust_remote_code=True,
    padding_side="right",
)

model = LlavaPythiaForCausalLM.from_pretrained(
    MODEL_DIR,
    config=config,
    trust_remote_code=True,
)

model.tie_weights()

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device).eval()

def chat_once(query: str, max_new_tokens: int = 128) -> str:
    conv = conversation_lib.default_conversation.copy()
    conv.append_message(conv.roles[0], query)
    prompt = conv.get_prompt()

    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    generated = inputs["input_ids"]

    for _ in range(max_new_tokens):
        outputs = model.gpt_neox(
            input_ids=generated,
            attention_mask=torch.ones_like(generated),
            return_dict=True,
        )
        hidden = outputs.last_hidden_state

        # 标准流程：final_layer_norm + embed_in 作为 lm_head
        normed = model.gpt_neox.final_layer_norm(hidden)
        lm_w = model.gpt_neox.embed_in.weight
        logits = normed[:, -1:, :] @ lm_w.T
        next_token = logits.argmax(dim=-1)

        generated = torch.cat([generated, next_token], dim=1)
        if next_token.item() == tokenizer.eos_token_id:
            break

    out_ids = generated[0, inputs["input_ids"].shape[1]:].tolist()
    return tokenizer.decode(out_ids, skip_special_tokens=True).strip()

if __name__ == "__main__":
    print("用户: how can i use knife")
    print("助手:", chat_once("how can i use knife"))
`

**Comment by zxczcggghkk** [comment]
good question

**Comment by GaryOAO** [comment]
Hi there! I have the same qustion as you, and I rerote a test python program to do some Q&A tasks and it runs well.
You may test Llava-Pythia-1.3B by using the following code :

```python
import torch
import torch.nn.functional as F
from transformers import AutoTokenizer
from llava_pythia.model.language_model.pythia.configuration_llava_pythia import LlavaPythiaConfig
from llava_pythia.model.language_model.pythia.llava_pythia import LlavaPythiaForCausalLM

def top_k_top_p_filtering(logits, top_k=0, top_p=0.0, filter_value=-float('Inf')):
    top_k = min(top_k, logits.size(-1))
    if top_k > 0:
        kth_val = torch.topk(logits, top_k)[0][..., -1, None]
        logits = torch.where(logits < kth_val, filter_value, logits)
    if top_p > 0.0:
        sorted_logits, sorted_idx = torch.sort(logits, descending=True)
        cum_probs = torch.cumsum(F.softmax(sorted_logits, dim=-1), dim=-1)
        mask = cum_probs > top_p
        mask[..., 1:] = mask[..., :-1].clone()
        mask[..., 0] = False
        remove_idx = sorted_idx[mask]
        logits[remove_idx] = filter_value
    return logits

def main():
    model_dir = "/YOUR_MODEL_PATH_HERE"
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")

    tokenizer = AutoTokenizer.from_pretrained(model_dir, use_fast=True, trust_remote_code=True)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    config = LlavaPythiaConfig.from_pretrained(model_dir, trust_remote_code=True)
    config.use_cache = True
    config.is_decoder = True
    setattr(config, 'action_head_type', 'none')
    setattr(config, 'action_dim', 0)

    model = LlavaPythiaForCausalLM.from_pretrained(
        model_dir,
        config=config,
        torch_dtype=torch.float32,
        trust_remote_code=True,
    )
    model.to(device).eval()

    max_new_tokens = 100
    temperature = 0.7
    top_k = 50
    top_p = 0.95

    question = "Who are you? What can you do?"
    prompt = f"Question: {question}\nAnswer:"
    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    generated = inputs.input_ids[0].tolist()

    for _ in range(max_new_tokens):
        curr_ids = torch.tensor([generated], device=device)
        curr_attn = torch.ones_like(curr_ids)

        with torch.no_grad():
            outputs = model(
                input_ids=curr_ids,
                attention_mask=curr_attn,
                return_dict=True
            )
            logits_all = outputs.logits

        next_logits = logits_all[0, -1, :] / temperature
        filtered = top_k_top_p_filtering(next_logits, top_k=top_k, top_p=top_p)
        probs = F.softmax(filtered, dim=-1)
        next_token = torch.multinomial(probs, 1).item()
        if next_token == tokenizer.eos_token_id:
            break
        generated.append(next_token)

    new_tokens = generated[len(inputs.input_ids[0]):]
    answer = tokenizer.decode(new_tokens, skip_special_tokens=True)
    print("Answer：", answer)

if __name__ == "__main__":
    main()


```

An exaple output is that : 
```bash
[2025-07-17 15:31:22,792] [INFO] [real_accelerator.py:110:get_accelerator] Setting ds_accelerator to cuda (auto detect)
Using device: cuda
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.
Answer：  I'm a computer program
```
Happy Coding!  👨‍💻
